File: app/src/main/java/de/kuschku/libquassel/primitives/types/BufferInfo.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.primitives.types;

import android.support.annotation.NonNull;

import com.raizlabs.android.dbflow.annotation.Column;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.structure.BaseModel;

import de.kuschku.libquassel.localtypes.orm.ConnectedDatabase;

@Table(database = ConnectedDatabase.class)
public class BufferInfo extends BaseModel {
    @PrimaryKey
    public int id;

    @Column
    public int networkId;

    @Column
    public Type type;

    @Column
    public int groupId;

    @Column
    public String name;

    public static BufferInfo create(int id, int networkId, Type type, int groupId, String name) {
        BufferInfo info = new BufferInfo();
        info.id = id;
        info.networkId = networkId;
        info.type = type;
        info.groupId = groupId;
        info.name = name;
        return info;
    }

    @NonNull
    @Override
    public String toString() {
        return "BufferInfo{" +
                "id=" + id +
                ", networkId=" + networkId +
                ", type=" + type +
                ", groupId=" + groupId +
                ", name='" + name + '\'' +
                '}';
    }

    public enum Type {
        INVALID(0x00),
        STATUS(0x01),
        CHANNEL(0x02),
        QUERY(0x04),
        GROUP(0x08);

        public final short id;

        Type(short id) {
            this.id = id;
        }

        Type(int id) {
            this((short) id);
        }

        @NonNull
        public static Type fromId(short id) {
            switch (id) {
                case 0x01:
                    return STATUS;
                case 0x02:
                    return CHANNEL;
                case 0x04:
                    return QUERY;
                case 0x08:
                    return GROUP;
                default:
                    return INVALID;
            }
        }

    }

    public enum BufferStatus {
        OFFLINE,
        AWAY,
        ONLINE
    }

    public enum ContentStatus {
        NONE,
        ACTIVITY,
        MESSAGES,
        HIGHLIGHTS
    }

}

File: lib/src/main/java/de/kuschku/libquassel/quassel/BufferInfo.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2019 Janne Koschinski
 * Copyright (c) 2019 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.quassel

import de.kuschku.libquassel.protocol.Buffer_Type
import de.kuschku.libquassel.protocol.Buffer_Types
import de.kuschku.libquassel.util.flag.Flag
import de.kuschku.libquassel.util.flag.Flags
import de.kuschku.libquassel.util.flag.ShortFlag
import de.kuschku.libquassel.util.flag.ShortFlags

data class BufferInfo(
  var bufferId: Int = -1,
  var networkId: Int = -1,
  var type: Buffer_Types = Buffer_Type.of(),
  var groupId: Int = -1,
  var bufferName: String? = null
) {
  enum class Type(override val bit: UShort) : ShortFlag<Type> {
    InvalidBuffer(0x00u),
    StatusBuffer(0x01u),
    ChannelBuffer(0x02u),
    QueryBuffer(0x04u),
    GroupBuffer(0x08u);

    companion object : ShortFlags.Factory<Type> {
      override val NONE = Buffer_Type.of()
      val validValues = values().filter { it.bit != 0u.toUShort() }.toTypedArray()
      override fun of(bit: Short) = ShortFlags.of<Type>(bit)
      override fun of(bit: UShort) = ShortFlags.of<Type>(bit)
      override fun of(vararg flags: Type) = ShortFlags.of(*flags)
      override fun of(flags: Iterable<Type>) = ShortFlags.of(flags)
    }
  }

  enum class Activity(override val bit: UInt) : Flag<Activity> {
    NoActivity(0x00u),
    OtherActivity(0x01u),
    NewMessage(0x02u),
    Highlight(0x04u);

    companion object : Flags.Factory<Activity> {
      override val NONE = Activity.of()
      override fun of(bit: Int) = Flags.of<Activity>(bit)
      override fun of(bit: UInt) = Flags.of<Activity>(bit)
      override fun of(vararg flags: Activity) = Flags.of(*flags)
      override fun of(flags: Iterable<Activity>) = Flags.of(flags)
    }
  }
}
